### PR TITLE
Use cluster-level index for failed watches

### PR DESCRIPTION
**Pull request by kargakis**
Ref https://github.com/openshift/origin/issues/11617

@liggitt any ideas how to test this

### CODE REVIEW

**Comment by kargakis** [comment]
cc: @mfojtik @pweil-

**Comment by pweil-** [comment]
note: I moved the PR to 1.5 milestone based on the soak time comment.  If this does turn out to be a blocker we can adjust.  Marking this PR as 1.5 as well (for now).

File: pkg/util/leaderlease/leaderlease.go [hunk 1]
```diff
@@ -128,10 +129,16 @@ func (e *Etcd) tryAcquire() (ok bool, ttl uint64, nextIndex uint64, err error) {
 
 	if latest.Node.Value != e.value {
 		glog.V(4).Infof("Lease %s owned by %s at %d ttl %d seconds, waiting for expiration", e.key, latest.Node.Value, nextIndex-1, ttl)
```
Comment (kargakis) [inline pkg/util/leaderlease/leaderlease.go]: Moved the fix in `eventIndexFor` but I still need to test it:)

File: pkg/util/leaderlease/leaderlease.go [hunk 2]
```diff
@@ -292,12 +296,14 @@ func (e *Etcd) waitExpiration(held bool, from uint64, stop chan struct{}) (bool,
 }
 
 // eventIndexFor returns the next etcd index to watch based on a response
-func eventIndexFor(resp *etcdclient.Response) uint64 {
-	if resp.Node != nil {
-		return resp.Node.ModifiedIndex + 1
-	}
-	if resp.PrevNode != nil {
-		return resp.PrevNode.ModifiedIndex + 1
+func eventIndexFor(resp *etcdclient.Response, useClusterLevelIndex bool) uint64 {
```
Comment (liggitt) [inline pkg/util/leaderlease/leaderlease.go]: If we are using the returned value to decide what to watch from, we should always use the cluster level index

**Comment by liggitt** [comment]
To test, create a lease long enough to let you flood etcd with unrelated changes and expire the watch window, then make sure the lease watch is able to be established again instead of flipping to polling mode

I know there are tests for the lease stuff, not sure if it's possible to drive this scenario in those, but that would be ideal

**Comment by liggitt** [comment]
I expected the fix to just change eventIndexFor to always return `resp.Index`

**Comment by kargakis** [comment]
@php-coder is taking this over

**Pull request closed without merging**